content/blog/write-git-commit-messages-first.md

```diff
@@ -0,0 +1,66 @@
+
+Recursively repeat this process with the new commit message as your latest goal until you arrive at work you can commit using the message you wrote beforehand. Unwind to the previously recorded commit message as the next goal until you arrive at delivering the original feature, but now with the minimal necessary changes. Let’s sketch out a short example:
+
+We pick up a story from the top of our backlog that asks us to show the user relevant data. We record our commit message first:  
+
+> “Display relevant data to the user.”
+
+As we attempt to figure out where we should make our changes, we find ourselves in a legacy application missing many tests. We then discover a file named apiClient.js, the responsibility of which is to know where and how to fetch data to be displayed to the user. It would appear that our changes would include swapping out the URL path in order to fetch the relevant data.
+
+```javascript
+import get from “http”;
+
+export default () = ({
+	execute: async () => 
+	{
+	    const data = await get(“/path/to/irrelevant/data”);
+		// ...
+	}
+});
+```
+
+In order to effectively make changes to this code, we’ll need to back it with tests. So we put aside our current task and write a new commit message that reflects the desired state of the code.
+
+> “Backfill apiClient.js missing tests.”
+
+As we attempt to write unit tests, we find that it’s difficult due to the statically imported HTTP client. We decide that the production code would benefit from having the HTTP client parameterized, so we stash our changes and write down the following new commit message:
+
+> “Parameterize http client from client.js.”
+
+At this point, our commit messages would look something like:
+
+```
+parameterize http client from client.js
+ └─┬ backfill apiClient.js missing tests
+   └─┬ display relevant data to the user
+```
+
+We are able to complete our latest goal using an IDE shortcut, commit using the recorded message, and push the commit independently of our other work. We pick up from the previous commit message and repeat this process until we’ve finished our feature.
```

Comment:
> We are able to complete our latest goal using an IDE shortcut

I might be misreading this, but is the "latest goal" mentioned here parameterizing the HTTP client, or stashing the commits and creating the new one? This just reads a bit ambiguously to me.